src/main/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporter.java

```diff
@@ -0,0 +1,305 @@
+            final Integer height = banner.getFormat().get(0).getH();
+
+            final ExtBanner extBanner = ExtBanner.builder()
+                    .sizes(
+                            width != null && height != null ? Arrays.asList(
+                                    Arrays.asList(width, height),
+                                    Arrays.asList(width, height)
+                            ) : null
+                    )
+                    .pos(banner.getPos())
+                    .name(banner.getId())
+                    .build();
+
+            final MediaTypes mediaTypes = MediaTypes.builder()
+                    .banner(extBanner)
+                    .video(video)
+                    .nativeObject(nativeObject)
+                    .build();
+
+            final List<GreenbidsBidder> bidders = new ArrayList<>();
+
+            final Map<String, Bid> seatsWithBidsForImp = seatsWithBids.entrySet().stream()
+                    .filter(entry -> entry.getValue().getImpid().equals(imp.getId()))
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+            final Map<String, NonBid> seatsWithNonBidsForImp = seatsWithNonBids.entrySet().stream()
+                    .filter(entry -> entry.getValue().getImpId().equals(imp.getId()))
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+            seatsWithBidsForImp.forEach((seat, bid) -> {
+                final GreenbidsBidder bidder = GreenbidsBidder.builder()
+                        .bidder(seat)
+                        .isTimeout(false)
+                        .hasBid(bid != null)
+                        .build();
+                bidders.add(bidder);
+            });
+
+            seatsWithNonBidsForImp.forEach((seat, nonBid) -> {
+                final GreenbidsBidder bidder = GreenbidsBidder.builder()
+                        .bidder(seat)
+                        .isTimeout(nonBid.getStatusCode().code == BidRejectionReason.TIMED_OUT.code)
+                        .hasBid(false)
+                        .build();
+                bidders.add(bidder);
+            });
+
+            final String adUnitCode = getAdUnitCode(imp);
+
+            return AdUnit.builder()
+                    .code(adUnitCode)
+                    .mediaTypes(mediaTypes)
+                    .bidders(bidders)
+                    .build();
+        }).toList();
+    }
+
+    private String getAdUnitCode(Imp imp) {
+        final ObjectNode impExt = imp.getExt();
+
+        final String adUnitCodeStoredRequestId = Optional.ofNullable(getGpid(impExt))
+                .orElseGet(() -> storedRequestId(impExt));
+
+        if (adUnitCodeStoredRequestId != null) {
+            return adUnitCodeStoredRequestId;
+        }
+
+        return imp.getId();
+    }
+
+    private static String getGpid(ObjectNode impExt) {
+        return Optional.ofNullable(impExt)
```

Comment:
Here we can actually indeed us an Optional chain to fallback gpId -> storedRequestId -> impId in `getAdUnitCode` as follows

```java
private String getAdUnitCode(Imp imp) {
        final ObjectNode impExt = imp.getExt();

        return Optional.ofNullable(getGpid(impExt))
                .or(() -> Optional.ofNullable(storedRequestId(impExt)))
                .orElse(imp.getId());
    }
```

then `getGpid` and `storedRequestId` are both the optional chains as well, so I propose to keep them as is and modify `getAdUnitCode` method. WDYT?

P.S. thanks for comment on this methods because I've fixed an error when extracting `getGpid`: extract the field using `gpid` key from ObjectNode as like it is done in [RubiconBidder](https://github.com/EvgeniiMunin/prebid-server-java/blob/35f990b5c35b4b571d90280f73d8e5191ef17d49/src/main/java/org/prebid/server/bidder/rubicon/RubiconBidder.java#L970)

Reply:
P.S. thanks for comment on this methods because I've fixed an error when extracting `getGpid`: extract the field using `gpid` key from ObjectNode as like it is done in [RubiconBidder](https://github.com/EvgeniiMunin/prebid-server-java/blob/35f990b5c35b4b571d90280f73d8e5191ef17d49/src/main/java/org/prebid/server/bidder/rubicon/RubiconBidder.java#L970)